lib/signs_ui/alerts/display.ex

```diff
@@ -0,0 +1,64 @@
+defmodule SignsUi.Alerts.Display do
+  alias SignsUi.Alerts.Alert
+  alias SignsUi.Alerts.State
+
+  @typep alert_map :: %{Alert.id() => Alert.t()}
+  @typep route_alerts_map :: %{Alert.route_id() => alert_map()}
+
+  defstruct alerts: %{}
+
+  @typep t :: %__MODULE__{
+          alerts: route_alerts_map()
+        }
+
+  @spec format_state(State.t()) :: t()
+  def format_state(state) do
+    %__MODULE__{
+      # Take every alert in the current state,
+      alerts:
+        state
+        # generate a list of single_route_alerts from it, and flatten,
+        |> Stream.flat_map(&expand_routes/1)
+        # group them by route,
+        |> Enum.group_by(fn alert -> alert.route end)
+        # convert each group's list of alerts into a map,
+        |> Stream.map(&alert_map_for_route/1)
+        # and convert the list of alert groups into a map.
+        |> Map.new()
+    }
+  end
+
+  @spec expand_routes({Alert.id(), Alert.multi_route()}) :: [Alert.single_route()]
+  defp expand_routes({id, alert}) do
+    # Take a multi_route_alert(), maps over its affected_routes, and create a
+    # single_route_alert() for every affected route.
+    alert.affected_routes
+    |> Enum.map(fn route ->
```

Comment:
Ah, made me realize we don't have `credo` set up for signs_ui, but we generally don't pipe just once, instead preferring here `Enum.map(alert.affected_routes, fn route -> ...`.

Reply:
Good point.